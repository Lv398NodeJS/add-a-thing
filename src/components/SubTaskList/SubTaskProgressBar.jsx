import React, { Component } from 'react';
import { Row, Col, ProgressBar } from 'react-bootstrap';

class SubTaskAdd extends Component {
  render() {
    const { subtaskList } = this.props;
    const completedSubTasks = subtaskList.filter(subTask => subTask.completed);
    const completedSubTasksPercent = (completedSubTasks.length / subtaskList.length) * 100;
    const progressBarVariant = (percent) => {
      switch (true) {
        case percent === 100:
          return 'success';
        case percent > 50:
          return 'info';
        case percent > 30:
          return 'warning';
        default:
          return 'secondary';
      }
    };
    return (
      <Row className="justify-content-sm-center">
        <Col>
          <ProgressBar
            className="mb-3"
            animated
            now={completedSubTasksPercent}
            variant={progressBarVariant(completedSubTasksPercent)}
          />
        </Col>
      </Row>
    );
  }
}

export default SubTaskAdd;
