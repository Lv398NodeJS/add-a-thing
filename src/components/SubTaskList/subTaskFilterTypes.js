export const subtaskFilterTypes = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE',
};
