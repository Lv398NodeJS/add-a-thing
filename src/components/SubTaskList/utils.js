export const getSubtaskListAsArray = (snapValue = {}) => {
  const subtaskList = [];
  Object.keys(snapValue).forEach((subtask) => {
    const { text, completed } = snapValue[subtask];
    subtaskList.push({
      text,
      completed,
      id: subtask,
    });
  });
  return subtaskList;
};
