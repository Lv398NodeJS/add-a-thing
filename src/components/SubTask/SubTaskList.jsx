/* eslint-disable no-param-reassign */
import React, { Component } from 'react';
import { Container, Row, Col } from 'react-bootstrap';
import './index.scss';
import uuidv1 from 'uuid/v1';
import SubTaskItem from './SubTaskItem';
import SubTaskAdd from './SubTaskAdd';
import SubTaskProgressBar from './SubTaskProgressBar';

class SubTaskList extends Component {
  constructor(props) {
    super(props);

    const { subtasks } = this.props;
    this.state = { subTasks: subtasks };

    this.changeSubTaskStatus = this.changeSubTaskStatus.bind(this);
    this.addSubTask = this.addSubTask.bind(this);
    this.deleteSubTask = this.deleteSubTask.bind(this);
  }

  changeSubTaskStatus(id) {
    this.setState((prevState) => {
      const updatedSubTasks = prevState.subTasks.map((subTask) => {
        if (subTask.id === id) subTask.completed = !subTask.completed;
        return subTask;
      });
      return { subTasks: updatedSubTasks };
    });
  }

  addSubTask(subTaskText) {
    subTaskText = subTaskText.trim();
    this.setState(prevState => ({
      subTasks: [...prevState.subTasks, { id: uuidv1(), completed: false, text: subTaskText }],
    }));
  }

  deleteSubTask(subTaskId) {
    this.setState((prevState) => {
      const updatedSubTasks = prevState.subTasks.filter(subTask => subTask.id !== subTaskId);
      return { subTasks: updatedSubTasks };
    });
  }

  render() {
    const { subTasks } = this.state;
    const subTaskItems = subTasks.map(subTask => (
      <SubTaskItem
        key={subTask.id}
        subTask={subTask}
        changeSubTaskStatus={this.changeSubTaskStatus}
        deleteSubTask={this.deleteSubTask}
      />
    ));

    return (
      <Container>
        <SubTaskProgressBar subTasks={subTasks} />
        {subTaskItems}
        <Row className="justify-content-sm-center">
          <Col>
            <SubTaskAdd addSubTask={this.addSubTask} />
          </Col>
        </Row>
      </Container>
    );
  }
}

export default SubTaskList;
