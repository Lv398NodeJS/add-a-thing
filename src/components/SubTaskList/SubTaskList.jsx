import React, { Component } from 'react';
import { Row, Col } from 'react-bootstrap';

import SubTaskItem from './SubTaskItem/SubTaskItem';
import SubTaskAdd from './SubTaskAdd/SubTaskAdd';
import SubTaskProgressBar from './SubTaskProgressBar/SubTaskProgressBar';

import getSubtaskListAsArray from './getSubtaskListAsArray';

export default class SubTaskList extends Component {
  constructor(props) {
    super(props);
    this.taskRef = props.taskRef;
    this.state = {
      taskStatus: '',
      subtaskList: [],
    };
  }

  componentDidMount() {
    this.isComponentMounted = true;
    if (!this.taskRef) return;
    this.taskRef.on('value', (snapshot) => {
      const { status, subtaskList = {} } = snapshot.val() || {};
      if (this.isComponentMounted) {
        this.setState({
          taskStatus: status,
          subtaskList: getSubtaskListAsArray(subtaskList),
        });
      }
    });
  }

  componentWillUnmount() {
    this.isComponentMounted = false;
  }

  addSubTask = (subTaskText) => {
    if (!subTaskText.trim().length) return;
    const subtaskListRef = this.taskRef.child('/subtaskList');
    subtaskListRef.push({ text: subTaskText.trim(), completed: false });
  };

  deleteSubTask = (subtaskId) => {
    const subtaskRef = this.taskRef.child(`/subtaskList/${subtaskId}`);
    subtaskRef.remove();
  };

  convertToTask = (subtaskId, text) => {
    const subtaskRef = this.taskRef.child(`/subtaskList/${subtaskId}`);
    subtaskRef.remove().then(
      this.taskRef.parent.push({
        name: text,
        description: '',
        status: 'To Do',
        priority: 'Low',
      }),
    );
  };

  changeSubTaskStatus = (subtaskId) => {
    const subtaskRef = this.taskRef.child(`/subtaskList/${subtaskId}`);
    subtaskRef.once('value', (snapshot) => {
      subtaskRef.set({
        text: snapshot.val().text,
        completed: !snapshot.val().completed,
      });
    });
  };

  render() {
    const { taskStatus, subtaskList } = this.state;
    const subTaskItems = subtaskList.map(subTask => (
      <SubTaskItem
        key={subTask.id}
        id={subTask.id}
        text={subTask.text}
        completed={subTask.completed}
        taskStatus={taskStatus}
        changeSubTaskStatus={this.changeSubTaskStatus}
        deleteSubTask={this.deleteSubTask}
        convertToTask={this.convertToTask}
      />
    ));

    return (
      <>
        <SubTaskProgressBar subtaskList={subtaskList} />
        {subTaskItems}
        <Row className="justify-content-sm-center">
          <Col>
            <SubTaskAdd taskStatus={taskStatus} addSubTask={this.addSubTask} />
          </Col>
        </Row>
      </>
    );
  }
}
