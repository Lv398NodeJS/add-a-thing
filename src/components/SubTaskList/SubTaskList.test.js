import React from 'react';
import { shallow } from 'enzyme';
import SubTaskList from './SubTaskList';

const subTaskListComponent = shallow(<SubTaskList />);
subTaskListComponent.setState({
  subtaskList: [
    { completed: false, id: '1', text: 'Test subtask 1' },
    { completed: false, id: '2', text: 'Test subtask 2' },
    { completed: false, id: '3', text: 'Test subtask 3' },
  ],
});

describe('SubTaskList component', () => {
  it('should render correctly and match the snapshot', () => {
    expect(subTaskListComponent).toMatchSnapshot();
  });

  it('should render SubTaskProgressBar child component', () => {
    expect(subTaskListComponent.find('SubTaskProgressBar').length).toBe(1);
  });

  it('should render SubTaskItem component for each sub task object in subtaskList array in state', () => {
    expect(subTaskListComponent.find('SubTaskItem').length).toBe(3);
  });

  it('should render SubTaskAdd child component', () => {
    expect(subTaskListComponent.find('SubTaskAdd').length).toBe(1);
  });
});
