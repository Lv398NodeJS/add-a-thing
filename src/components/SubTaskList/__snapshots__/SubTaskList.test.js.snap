// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubTaskList component should render correctly and match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SubTaskList />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "as": "div",
      "children": Array [
        <SubTaskProgressBar
          subtaskList={
            Array [
              Object {
                "completed": false,
                "id": "1",
                "text": "Test subtask 1",
              },
              Object {
                "completed": false,
                "id": "2",
                "text": "Test subtask 2",
              },
              Object {
                "completed": false,
                "id": "3",
                "text": "Test subtask 3",
              },
            ]
          }
        />,
        Array [
          <SubTaskItem
            changeSubTaskStatus={[Function]}
            completed={false}
            deleteSubTask={[Function]}
            id="1"
            text="Test subtask 1"
          />,
          <SubTaskItem
            changeSubTaskStatus={[Function]}
            completed={false}
            deleteSubTask={[Function]}
            id="2"
            text="Test subtask 2"
          />,
          <SubTaskItem
            changeSubTaskStatus={[Function]}
            completed={false}
            deleteSubTask={[Function]}
            id="3"
            text="Test subtask 3"
          />,
        ],
        <ForwardRef(render)
          className="justify-content-sm-center"
        >
          <ForwardRef
            as="div"
          >
            <SubTaskAdd
              addSubTask={[Function]}
            />
          </ForwardRef>
        </ForwardRef(render)>,
      ],
      "fluid": false,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "subtaskList": Array [
            Object {
              "completed": false,
              "id": "1",
              "text": "Test subtask 1",
            },
            Object {
              "completed": false,
              "id": "2",
              "text": "Test subtask 2",
            },
            Object {
              "completed": false,
              "id": "3",
              "text": "Test subtask 3",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "changeSubTaskStatus": [Function],
          "completed": false,
          "deleteSubTask": [Function],
          "id": "1",
          "text": "Test subtask 1",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "changeSubTaskStatus": [Function],
          "completed": false,
          "deleteSubTask": [Function],
          "id": "2",
          "text": "Test subtask 2",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "class",
        "props": Object {
          "changeSubTaskStatus": [Function],
          "completed": false,
          "deleteSubTask": [Function],
          "id": "3",
          "text": "Test subtask 3",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ForwardRef
            as="div"
          >
            <SubTaskAdd
              addSubTask={[Function]}
            />
          </ForwardRef>,
          "className": "justify-content-sm-center",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "div",
            "children": <SubTaskAdd
              addSubTask={[Function]}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "addSubTask": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "as": "div",
            },
            "displayName": "Col",
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": undefined,
          "displayName": "Bootstrap(Row)",
          "propTypes": undefined,
          "render": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "as": "div",
        "fluid": false,
      },
      "displayName": "Container",
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "as": "div",
        "children": Array [
          <SubTaskProgressBar
            subtaskList={
              Array [
                Object {
                  "completed": false,
                  "id": "1",
                  "text": "Test subtask 1",
                },
                Object {
                  "completed": false,
                  "id": "2",
                  "text": "Test subtask 2",
                },
                Object {
                  "completed": false,
                  "id": "3",
                  "text": "Test subtask 3",
                },
              ]
            }
          />,
          Array [
            <SubTaskItem
              changeSubTaskStatus={[Function]}
              completed={false}
              deleteSubTask={[Function]}
              id="1"
              text="Test subtask 1"
            />,
            <SubTaskItem
              changeSubTaskStatus={[Function]}
              completed={false}
              deleteSubTask={[Function]}
              id="2"
              text="Test subtask 2"
            />,
            <SubTaskItem
              changeSubTaskStatus={[Function]}
              completed={false}
              deleteSubTask={[Function]}
              id="3"
              text="Test subtask 3"
            />,
          ],
          <ForwardRef(render)
            className="justify-content-sm-center"
          >
            <ForwardRef
              as="div"
            >
              <SubTaskAdd
                addSubTask={[Function]}
              />
            </ForwardRef>
          </ForwardRef(render)>,
        ],
        "fluid": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "subtaskList": Array [
              Object {
                "completed": false,
                "id": "1",
                "text": "Test subtask 1",
              },
              Object {
                "completed": false,
                "id": "2",
                "text": "Test subtask 2",
              },
              Object {
                "completed": false,
                "id": "3",
                "text": "Test subtask 3",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "changeSubTaskStatus": [Function],
            "completed": false,
            "deleteSubTask": [Function],
            "id": "1",
            "text": "Test subtask 1",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "changeSubTaskStatus": [Function],
            "completed": false,
            "deleteSubTask": [Function],
            "id": "2",
            "text": "Test subtask 2",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "changeSubTaskStatus": [Function],
            "completed": false,
            "deleteSubTask": [Function],
            "id": "3",
            "text": "Test subtask 3",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ForwardRef
              as="div"
            >
              <SubTaskAdd
                addSubTask={[Function]}
              />
            </ForwardRef>,
            "className": "justify-content-sm-center",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": "div",
              "children": <SubTaskAdd
                addSubTask={[Function]}
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "addSubTask": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "as": "div",
              },
              "displayName": "Col",
              "render": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "displayName": "Bootstrap(Row)",
            "propTypes": undefined,
            "render": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "as": "div",
          "fluid": false,
        },
        "displayName": "Container",
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
